File: 1. Fundamental-Level/CSharp/03. OOP-November-2015/[HW]/04. Inheritance-And-Abstraction/04. [EX]/01. Book-Shop/Book.cs
using System;

public class Book
{
    private string title;
    private string author;
    private double price;

    public Book(string title, string author, double price)
    {
        this.Author = author;
        this.Title = title;
        this.Price = price;
    }

    public string Title
    {
        get
        {
            return this.title;
        }
        protected set
        {
            if (string.IsNullOrEmpty(value))
            {
                throw new ArgumentNullException("The title cannot be empty.");
            }

            this.title = value;
        }
    }
    public string Author
    {
        get
        {
            return this.author;
        }
        protected set
        {
            if (string.IsNullOrEmpty(value))
            {
                throw new ArgumentNullException("The title cannot be empty.");
            }

            this.author = value;
        }
    }
    public virtual double Price
    {
        get
        {
            return this.price;
        }
        protected set
        {
            if (value < 0)
            {
                throw new ArgumentOutOfRangeException("The price cannot be negative.");
            }

            this.price = value;
        }
    }

    public override string ToString()
    {
        return string.Format("-Type: {0}\n-Title: {1}\n-Author: {2}\n-Price: {3}", this.GetType(), this.title, this.Author, this.Price);
    }
}

File: 1. Fundamental-Level/Java/OOP-Bacis-June-2016/[EX]/05. Inheritance/src/problem2_BookShop/Book.java
package problem2_BookShop;

public class Book {
    private String title;
    private String author;
    private double price;

    public Book(String author, String title, double price) {
        this.setTitle(title);
        this.setAuthor(author);
        this.setPrice(price);
    }

    private void setTitle(String title) {
        if (title == null || title.trim().length() < 3){
            throw new IllegalArgumentException("Title not valid!");
        }

        this.title = title;
    }


    private void setAuthor(String author) {
        if(author == null || !this.isAuthorNameValid(author)){
            throw new IllegalArgumentException("Author not valid!");
        }

        this.author = author;
    }

    protected void setPrice(double price) {
        if (price <= 0){
            throw new IllegalArgumentException("Price not valid!");
        }

        this.price = price;
    }

    private boolean isAuthorNameValid(String author) {
        String[] authorParams = author.split("\\s+");
        if (authorParams.length > 1 && Character.isDigit(authorParams[1].charAt(0))){
            return false;
        }

        return true;
    }

    @Override
    public String toString() {
        return "Type: " + this.getClass().getSimpleName() + System.lineSeparator() +
                "Title: " + this.title + System.lineSeparator() +
                "Author: " + this.author + System.lineSeparator() +
                "Price: " + String.format("%.1f", this.price) + System.lineSeparator();
    }
}